homeassistant/components/homeassistant/triggers/event.py

```diff
@@ -47,38 +48,54 @@ async def async_attach_trigger(
     event_types = template.render_complex(
         config[CONF_EVENT_TYPE], variables, limited=True
     )
-    removes = []
-
-    event_data_schema = None
+    event_data_schema: vol.Schema | None = None
+    event_data_items: ItemsView | None = None
     if CONF_EVENT_DATA in config:
         # Render the schema input
         template.attach(hass, config[CONF_EVENT_DATA])
         event_data = {}
         event_data.update(
             template.render_complex(config[CONF_EVENT_DATA], variables, limited=True)
         )
-        # Build the schema
-        event_data_schema = vol.Schema(
-            {vol.Required(key): value for key, value in event_data.items()},
-            extra=vol.ALLOW_EXTRA,
-        )
-
-    event_context_schema = None
+        # Build the schema or a an items view if the schema is simple
+        # and does not contain sub-dicts. We explicitly do not check for
+        # list like the context data below since lists are a special case
+        # only for context data. (see test test_event_data_with_list)
+        if any(isinstance(value, dict) for value in event_data.values()):
```

Comment:
We actually don't check sub-dicts. If a key exists, the value needs to 1:1 match. I think that we can always use the `items()` check.

Reply:
Sadly it appears nested/sub-dicts are intended to be supported since #9732 added support for them as well and test_if_fires_on_event_with_nested_data fails without the fallback